File: src/main/java/at/raphaell/inference/Chunker.java
package at.raphaell.inference;

import at.raphaell.inference.models.Paper;
import java.util.ArrayList;
import java.util.List;

public class Chunker {

    //    TODO make var
    private static final int CHUNK_SIZE = 128;
    private static final int CHUNK_OVERLAP = 32;

    public static List<ChunksWithPaper> createChunks(final Paper paper) {
        final String inputText = paper.abstract_();
        final List<ChunksWithPaper> chunks = new ArrayList<>();


        int textLength = inputText.length();
        int start = 0;

        while (start < textLength) {
            int end = Math.min(start + CHUNK_SIZE, textLength);
            String chunk = inputText.substring(start, end);
            chunks.add(new ChunksWithPaper(chunk, paper));

            start += (CHUNK_SIZE - CHUNK_OVERLAP);
        }

        return chunks;
    }

    public record ChunksWithPaper(String chunk, Paper paper) {}
}

File: inference-app/src/chunker.py
from config import CHUNK_OVERLAP, CHUNK_SIZE
from paper import with_chunk, Paper


# TODO use tokens instead of char len!
def create_chunks(paper) -> list[Paper]:
    input_text = paper.abstract
    chunks = []
    text_length = len(input_text)
    start = 0

    while start < text_length:
        end = min(start + CHUNK_SIZE, text_length)
        chunk = input_text[start:end]
        chunks.append(with_chunk(paper, chunk))
        start += CHUNK_SIZE - CHUNK_OVERLAP

    return chunks
